libraries/AP_HAL_PX4/I2CDevice.cpp

```diff
@@ -0,0 +1,73 @@
+#include "I2CDevice.h"
+
+#include <AP_HAL/AP_HAL.h>
+
+namespace PX4 {
+
+class PX4_I2C : public device::I2C {
+public:
+    PX4_I2C(uint8_t bus);
+    bool do_transfer(uint8_t address, const uint8_t *send, uint32_t send_len,
+                     uint8_t *recv, uint32_t recv_len);
+};
+
+PX4_I2C::PX4_I2C(uint8_t bus) :
+    I2C("AP_I2C", "/dev/api2c", bus, 0, 400000UL)
+{
+    init();
+}
+
+bool PX4_I2C::do_transfer(uint8_t address, const uint8_t *send, uint32_t send_len,
+                          uint8_t *recv, uint32_t recv_len)
+{
+    set_address(address);
+    return transfer(send, send_len, recv, recv_len) == OK;
+}
+
+bool I2CDevice::transfer(const uint8_t *send, uint32_t send_len,
+                         uint8_t *recv, uint32_t recv_len)
+{
+    return _device.do_transfer(_address, send, send_len, recv, recv_len);
+}
+
+bool I2CDevice::read_registers_multiple(uint8_t first_reg, uint8_t *recv,
+                                        uint32_t recv_len, uint8_t times)
+{
+    while (times > 0) {
+        if (!transfer(nullptr, 0, recv, recv_len)) {
+            return false;
+        }
+
+        recv += recv_len;
+        times--;
+    }
+    return true;
+}
+
+int I2CDevice::get_fd()
+{
+    return -1;
+}
+
+AP_HAL::Semaphore *I2CDevice::get_semaphore()
+{
+    return &semaphore;
+}
+
+I2CDevice::~I2CDevice()
+{
+}
+
+I2CDeviceManager::I2CDeviceManager()
+{
+}
+
+AP_HAL::OwnPtr<AP_HAL::I2CDevice>
+I2CDeviceManager::get_device(uint8_t bus, uint8_t address)
+{
+    AP_HAL::OwnPtr<PX4_I2C> i2c { new PX4_I2C(bus) };
+    auto dev = AP_HAL::OwnPtr<AP_HAL::I2CDevice>(new I2CDevice(*i2c, address));
+
+    return dev;
```

Comment:
If you do this, i2c object (from PX4_I2C) will be destroyed as soon as this function returns, regardless if you passed a reference to I2CDevice. Here you need to either
  1. Change I2CDevice constructor to receive an AP_HAL::OwnPtr<PX4_I2C> and then here call it using std::move() or
  2. Just embed PX4_I2C in I2CDevice

(1) means you are passing the ownership of the new PX4_I2C object to the I2CDevice object.... this is more or less what the Linux implementation does, but sharing the fd.

Since this is a much simpler implementation I did a commit with (2) on top of yours.